File: cpp/largest-series-product/largest_series_product.cpp
#include "largest_series_product.h"

#include <algorithm>
#include <cstddef>
#include <stdexcept>

namespace largest_series_product {

/**
 * @brief Checks if a string contains only digit characters.
 *
 * @param number The string to check.
 * @return `true` if the string contains only digits, `false` otherwise.
 */
bool is_number(const std::string &number);
/**
 * @brief Calculates the product of all digits in a given string.
 *
 * @param number The string of digits.
 * @return The product of the digits.
 */
uint64_t digits_product(const std::string &number);

/**
 * @brief Finds the largest product of a contiguous series of digits in a number string.
 *
 * This function calculates the product of all contiguous series of a specified length
 * and returns the maximum product found. It validates the input string and span,
 * throwing an exception if they are invalid.
 *
 * @param number The string of digits to analyze.
 * @param span The length of the contiguous series.
 * @return The largest product of a series of digits of the given length.
 * @throws std::domain_error if `span` is negative, greater than the string's length,
 * or if the input string contains non-digit characters.
 */
uint64_t largest_product(const std::string &number, int64_t span) {
  if (span < 0) {
    throw std::domain_error("span must be positive");
  }
  if (static_cast<int64_t>(number.length()) < span) {
    throw std::domain_error("span cannot be greater than string");
  }
  if (!is_number(number)) {
    throw std::domain_error(
        "The provided string is not a valid number because it contains "
        "non-digit characters.");
  }
  const size_t limit = number.length() - static_cast<size_t>(span) + 1;
  uint64_t max = 0;

  for (size_t i = 0; i < limit; i++) {
    max = std::max(max, digits_product(number.substr(i, span)));
  }

  return max;
}

/**
 * @brief Checks if a string contains only digit characters.
 *
 * @param number The string to check.
 * @return `true` if the string contains only digits, `false` otherwise.
 */
bool is_number(const std::string &number) {
  return std::all_of(number.begin(), number.end(),
                     [](char chr) { return std::isdigit(chr); });
}

/**
 * @brief Calculates the product of all digits in a given string.
 *
 * @param number The string of digits.
 * @return The product of the digits.
 */
uint64_t digits_product(const std::string &number) {
  uint64_t product = 1;

  for (const char chr : number) {
    product *= (chr - '0');
  }

  return product;
}

}  // namespace largest_series_product

File: typescript/largest-series-product/largest-series-product.ts
/**
 * Finds the largest product of a contiguous series of digits of a given length within a number string.
 *
 * This function validates the input, extracts all contiguous slices of a specified length,
 * calculates the product of the digits for each slice, and returns the maximum product found.
 *
 * @param num The string containing the digits.
 * @param span The length of the series of digits to consider.
 * @returns The largest product found.
 * @throws Error If `span` is negative, `span` is larger than the number string length, or the input string contains non-digit characters.
 *
 * @example
 * largestProduct('576802143', 2); // Returns 48 (from the slice '68')
 * largestProduct('1027839564', 3); // Returns 270 (from the slice '956')
 */
export function largestProduct(num: string, span: number): number {
  if (span < 0) {
    throw new Error("Span must not be negative");
  }
  if (span > num.length) {
    throw new Error("Span must be smaller than string length");
  }
  if (!isNumber(num)) {
    throw new Error("Digits input must only contain digits");
  }

  const multiplications = Array
    .from(slices(num, span))
    .map(digitsMultiplication);

  return Math.max(...multiplications);
}

/**
 * A generator function that produces contiguous slices of a string.
 *
 * It iterates through the `source` string and yields substrings of length `slicesLen`,
 * moving one character at a time.
 *
 * @param source The string to be sliced.
 * @param slicesLen The length of each slice.
 * @yields A string slice.
 */
function* slices(source: string, slicesLen: number): Generator<string, void, void> {
  const topLimit = source.length - slicesLen + 1;

  for (let i = 0; i < topLimit; i++) {
    yield source.slice(i, i + slicesLen);
  }
}

/**
 * Calculates the product of all digits in a string.
 *
 * The string is first split into individual characters, each is converted to a number,
 * and then all numbers are multiplied together.
 *
 * @param num The string of digits.
 * @returns The product of the digits.
 *
 * @example
 * digitsMultiplication('123'); // Returns 6
 * digitsMultiplication('576'); // Returns 210
 */
function digitsMultiplication(num: string): number {
  return num
    .split("")
    .map(Number)
    .reduce((prev, val) => prev * val);
}

/**
 * Checks if a string contains only digits (0-9).
 *
 * @param num The string to check.
 * @returns `true` if the string contains only digits, `false` otherwise.
 */
function isNumber(num: string): boolean {
  return !/[^0-9]/.test(num)
}